src/components/TradeWidget/index.tsx

```diff
@@ -361,7 +361,13 @@ const TradeWidget: React.FC = () => {
   // Avoid displaying an empty list of tokens when the wallet is not connected
   const fallBackNetworkId = networkId ? networkId : Network.Mainnet // fallback to mainnet
 
-  const tokens = useMemo(() => tokenListApi.getTokens(fallBackNetworkId), [fallBackNetworkId])
+  const { balances } = useTokenBalances()
+
+  // If user is connected, use balances, otherwise get the default list
+  const tokens = useMemo(
+    () => (isConnected && balances.length > 0 ? balances : tokenListApi.getTokens(fallBackNetworkId)),
```

Comment:
I find strange how we return either balances or the list of tokens
Wouldn't it be better to have it as two separate state properties, each one with it's own type, and then depending on `isConnected && balances.length` we render one or the other one?

Reply:
We could yes, and we had it before. I'm changing the behaviour because we'll only need the list of tokens when logged out.
I could argue that we are saving one API call to tokenListApi, although that's a local call.

Still, if you feel strongly about this I can revert.